core/src/main/java/org/elasticsearch/search/fetch/percolator/HighlightSubFetchPhase.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BoostQuery;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.text.Text;
+import org.elasticsearch.index.percolator.PercolatorFieldMapper;
+import org.elasticsearch.index.percolator.PercolatorQueryCache;
+import org.elasticsearch.index.query.ParsedQuery;
+import org.elasticsearch.index.query.PercolatorQuery;
+import org.elasticsearch.search.SearchParseElement;
+import org.elasticsearch.search.fetch.FetchSubPhase;
+import org.elasticsearch.search.highlight.HighlightPhase;
+import org.elasticsearch.search.highlight.SearchContextHighlight;
+import org.elasticsearch.search.internal.InternalSearchHit;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.internal.SubSearchContext;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+// Highlighting in the case of the percolator query is a bit different, because the PercolatorQuery itself doesn't get highlighted,
+// but the source of the PercolatorQuery gets highlighted by each hit with type '.percolator' (percolator queries).
+public class HighlightSubFetchPhase implements FetchSubPhase {
+
+    private final HighlightPhase highlightPhase;
+
+    @Inject
+    public HighlightSubFetchPhase(HighlightPhase highlightPhase) {
+        this.highlightPhase = highlightPhase;
+    }
+
+    @Override
+    public boolean hitsExecutionNeeded(SearchContext context) {
+        return context.highlight() != null && locatePercolatorQuery(context.query()) != null;
+    }
+
+    @Override
+    public void hitsExecute(SearchContext context, InternalSearchHit[] hits) {
+        PercolatorQuery percolatorQuery = locatePercolatorQuery(context.query());
+        if (percolatorQuery == null) {
+            // shouldn't happen as we checked for the existence of a percolator query in hitsExecutionNeeded(...)
+            throw new IllegalStateException("couldn't locate percolator query");
+        }
+
+        List<LeafReaderContext> ctxs = context.searcher().getIndexReader().leaves();
+        PercolatorQueryCache queriesRegistry = context.percolatorQueryCache();
+        IndexSearcher percolatorIndexSearcher = percolatorQuery.getPercolatorIndexSearcher();
+
+        LeafReaderContext percolatorLeafReaderContext = percolatorIndexSearcher.getIndexReader().leaves().get(0);
+        FetchSubPhase.HitContext hitContext = new FetchSubPhase.HitContext();
+        SubSearchContext subSearchContext =
+                createSubSearchContext(context, percolatorLeafReaderContext, percolatorQuery.getDocumentSource());
+
+        for (InternalSearchHit hit : hits) {
+            if (PercolatorFieldMapper.TYPE_NAME.equals(hit.getType())) {
```

Comment:
it is only for highlighting percolator queries in the search API. The percolator api does nothing else then delegating to seach api.